Source/Core/Core/MemTools.cpp

```diff
@@ -60,11 +62,28 @@ static LONG NTAPI Handler(PEXCEPTION_POINTERS pPtrs)
     // virtual address of the inaccessible data
     uintptr_t fault_address = (uintptr_t)pPtrs->ExceptionRecord->ExceptionInformation[1];
     SContext* ctx = pPtrs->ContextRecord;
-
-    if (Core::System::GetInstance().GetJitInterface().HandleFault(fault_address, ctx))
+    Core::System& system = Core::System::GetInstance();
+    Memory::MemoryManager& memory = system.GetMemory();
+    if (system.GetJitInterface().HandleFault(fault_address, ctx))
     {
       return EXCEPTION_CONTINUE_EXECUTION;
     }
+    else if (!memory.IsPageDirty(fault_address))
+    {
+      memory.SetPageDirtyBit(fault_address, 1, true);
```

Comment:
The range you pass to `SetPageDirtyBit` should be the whole host page, not `fault_address` plus one byte.

Imagine the size of the host page is 64K, which is 16 times larger than the size of the guest page (4K). Let's say there's a write to the first byte of the host page. This will cause the whole host page to become unprotected, so you need to mark 16 guest pages at dirty, not just one.